src/hardware/iohandler.cpp

```diff
@@ -27,8 +27,8 @@
 
 //#define ENABLE_PORTLOG
 
-IO_WriteHandler * io_writehandlers[3][IO_MAX];
-IO_ReadHandler * io_readhandlers[3][IO_MAX];
+IO_WriteHandler io_writehandlers[3][IO_MAX];
+IO_ReadHandler io_readhandlers[3][IO_MAX];
```

Comment:
Uh, oh; I am happy about change to std::function in this PR in general, but this table is quite big…

AFAIK (maybe it's outdated knowledge already), size of `std::function` object is implementation-defined. For GCC on Linux it's 32 bytes vs 8 bytes for plain old function pointer.

So these two tables grow in size from 1MB to 4MB… :(

Maybe better let's go back to plain old function pointers for these two tables, at least for now.

Reply:
Gack! great catch.  I remember coming across this massive table many months ago when I was trying to reduce these gigantic arrays (I ran actual measurements, and the largest number of handles was well under 1000.. but the code blasts this into the stratosphere with `#define IO_MAX (64 * 1024 + 3)`!

Ugh.  This is going to require rolling back some nice stuff in the GUS code and going back to globals and free-standing static functions; but I agree that it's not worth blowing up our memory usage over.

I suspect there's an elegant solution involving a `vector` of these guys, that just grows on demand during the startup phase, and we still gain all the performance benefit of a contiguous random-access array. (A std::map would be the most memory efficient, but lookups would be too slow).

Reply:
Heh, if number of real function pointers stored in here is really so small, then a new task could replace these big tables with much smaller hashmaps… This could actually result in some speedups by reducing memory footprint and avoiding cache misses. Perhaps better addressed in a separate PR though.

Reply:
Broke this out into it's own PR, and dropped it from this one.